### PR TITLE
enforce `no_std` for all? protocol/v2 crates

### PR DESCRIPTION
**Pull request by Georges760**
Most of protocol/v2 crates are almost `no_std`, as listed [here](https://github.com/stratum-mining/stratum/issues/1130#issuecomment-2438642786).

Having a lib crate that can be `no_std` or `std` by the selection of a feature :
```rust
#![cfg_attr(not(feature = "std"), no_std)]
```
or the current implementation :
```rust
#![cfg_attr(feature = "no_std", no_std)]
```
is only usefull if :
- the crate could use some `std` related dependency (for instance `std::vec::Vec`)
- but we want the crate's user to be able to not have a dependancy on `std` so we disable the functionalities using `std`

In some case we can replace the `std` dep by its `alloc` equivalent if exists (`alloc::vec::Vec`) and we can keep the functionality in `no_std` environement.

Finally, a `std` binary can use some `no_std` dependencies, but an `no_std` binary cannot use any `std` deps. So making the low level crates all `no_std` will make them copamtible for every kind of binary (app).

We should use the `std` feature only if some functionality requires `std` and not have any `no_std` equivalent.

As far as I tried all protocol/v2/ crates (libs) can be 100% `no_std` a,d does not need the `std` (or `no_std`) feature to optionally add `std` dep.

This PR is targeting to get rid of `std` dep without loosing any functionality or efficiency.

Note: I will make 1 commit per crate to explain what and how I am doing it.
Note2: being `no_alloc` is another level of complexity for deep embedded system, which is not at all this PR concern.

### CODE REVIEW

File: protocols/DEPENDANCIES.md
Comment (Georges760) [reply protocols/DEPENDANCIES.md:1]: sure !

Comment (plebhash) [reply protocols/DEPENDANCIES.md:1]: thanks!

now that we have #1233 we should drop c658909741686ac6a9d9269057f534ae003f3ebb from this PR

Comment (Georges760) [reply protocols/DEPENDANCIES.md:1]: removed

**Review by jbesraa — COMMENTED** [review]
While I think the general idea is very welcome, I think addressing https://github.com/stratum-mining/stratum/issues/1200 first would make more sense as it can potentially change how we approach this. I think in #1200 we should aim to provide a logger for std and no std env and then `impl Debug` manually, which would allow us to remove the debug feature IIUC whats going here

**Comment by Georges760** [comment]
Current `binary_codec_sv2` :
- don't compile without `no_std` feature (this doesn't mean it is `no_std` as there is no enforcement => `#![cfg_attr(feature = "no_std", no_std)]`)
- has `no_std` feature by default (`default = ["no_std"]`)
- nowhere (not even in the CI), its default features are disable (`--no-default-features`)

So it's pseudo `std` support os broken and unused. We can :
- remove/clean it, and keep a real `no_std` crate only
- debug it, and have a `no_std` crate by default, with a `std` feature with the debugged `std` functionalities

Maybe this can be discussed in #1132 ?

**Comment by Georges760** [comment]
> While I think the general idea is very welcome, I think addressing #1200 first would make more sense as it can potentially change how we approach this. I think in #1200 we should aim to provide a logger for std and no std env and then `impl Debug` manually, which would allow us to remove the debug feature IIUC whats going here

`tracing` has a `std` feature enabled by default, this means it actually add `std` dep to the crate that use it without disabling default features : `sv1_api`, `binary_sv2`, `codec_sv2` and `roles_logic_sv2` (the 4 crates listed in #1200). 2 of this 4 crates are targeted by this PR (`binary_sv2` and `codec_sv2`) and in order for them to become `no_std`, I will have to `no-default-features` tracing, and enable `std` in tracing using a local feature too. This can be done before or after #1200 without problem.

If I do it before #1200, when #1200 will remove `tracing` with the custom logger, the `std` feature will be available (and can be removed if not used after all).
If #1200 should be done before, then we can limit this PR to crates not depending on `binary_sv2` and `codec_sv2`, and wait to see what the new logger will propose for `no_std` support.

**Comment by jbesraa** [comment]
> > While I think the general idea is very welcome, I think addressing #1200 first would make more sense as it can potentially change how we approach this. I think in #1200 we should aim to provide a logger for std and no std env and then `impl Debug` manually, which would allow us to remove the debug feature IIUC whats going here
> 
> `tracing` has a `std` feature enabled by default, this means it actually add `std` dep to the crate that use it without disabling default features : `sv1_api`, `binary_sv2`, `codec_sv2` and `roles_logic_sv2` (the 4 crates listed in #1200). 2 of this 4 crates are targeted by this PR (`binary_sv2` and `codec_sv2`) and in order for them to become `no_std`, I will have to `no-default-features` tracing, and enable `std` in tracing using a local feature too. This can be done before or after #1200 without problem.
> 
> If I do it before #1200, when #1200 will remove `tracing` with the custom logger, the `std` feature will be available (and can be removed if not used after all). If #1200 should be done before, then we can limit this PR to crates not depending on `binary_sv2` and `codec_sv2`, and wait to see what the new logger will propose for `no_std` support.

Yea I think #1200 shouldn't necessarily be about removing `tracing` crate, but generalizing how we use the logger 
and _mainly_ enabling a logger in a non-std env. Making everything non-std by default would make it a bit tricky to debug the crates generally.
With that said, If #1200 will require too much time and would block this for too long, I am happy for this to get in anyway.

**Comment by Georges760** [comment]
Actual `binary_sv2` without default feature (`core`) and with `with_serde` feature, test is broken.

This PR is not fixing this.

CI does not see it because does not test it.

File: protocols/v2/binary-sv2/binary-sv2/src/lib.rs
Comment (plebhash) [inline protocols/v2/binary-sv2/binary-sv2/src/lib.rs:45]: why are these `with_serde` flags being added?

File: protocols/v2/binary-sv2/binary-sv2/src/lib.rs
Comment (plebhash) [inline protocols/v2/binary-sv2/binary-sv2/src/lib.rs:3]: why aren't we following this pattern here?
```
#![cfg_attr(feature = "no_std", no_std)]
```

**Comment by plebhash** [comment]
@Georges760 please note that we're aiming to deprecate `with_serde` feature soon.

I cannot give you a precise timeline but Q1 2025 would be my rough guess.

I wonder if that would make this process simpler?

File: protocols/v2/binary-sv2/no-serde-sv2/derive_codec/src/lib.rs
Comment (plebhash) [inline protocols/v2/binary-sv2/no-serde-sv2/derive_codec/src/lib.rs:1]: why aren't we following this pattern here?
```
#![cfg_attr(feature = "no_std", no_std)]
```

File: protocols/v2/binary-sv2/serde-sv2/src/lib.rs
Comment (plebhash) [inline protocols/v2/binary-sv2/serde-sv2/src/lib.rs:75]: why aren't we following this pattern here?
```
#![cfg_attr(feature = "no_std", no_std)]
```

File: protocols/v2/binary-sv2/binary-sv2/src/lib.rs
Comment (Georges760) [reply protocols/v2/binary-sv2/binary-sv2/src/lib.rs:45]: Enforcing #![no_std] at the top of lib.rs remove the implicitly use of std::vec::Vec everywhere in the crate. So here in one of the test that is run only with feature "with_serde" the Vec must be explicitly used from alloc instead of std. Because the test is depends on with_serde feature, I made the "use alloc::vec::Vec" also depending on with_serde feature to avoid a warning if we run the test without it.

File: protocols/v2/binary-sv2/binary-sv2/src/lib.rs
Comment (Georges760) [reply protocols/v2/binary-sv2/binary-sv2/src/lib.rs:3]: Because this crate is no_std compliant no matter what. No need to use a feature for it. KISS

File: protocols/v2/binary-sv2/no-serde-sv2/derive_codec/src/lib.rs
Comment (Georges760) [reply protocols/v2/binary-sv2/no-serde-sv2/derive_codec/src/lib.rs:1]: Same: KISS.

File: protocols/v2/binary-sv2/serde-sv2/src/lib.rs
Comment (Georges760) [reply protocols/v2/binary-sv2/serde-sv2/src/lib.rs:75]: KISS.

**Comment by Georges760** [comment]
> @Georges760 please note that we're aiming to deprecate `with_serde` feature soon.
> 
> I cannot give you a precise timeline but Q1 2025 would be my rough guess.
> 
> I wonder if that would make this process simpler?

It make sense, as it is already broken in binary_sv2 and framing_sv2 as stated above. 

So I understand the official/only codec way will be the "core" option ? Good

**Comment by Georges760** [comment]
`noise_sv2` has its own PR #1238.
`codec_sv2` is dependant on both `noise_sv2` and all the crates of this PR.

So I stop here for this PR and open for reviews.
I think it is in good shape to fullfil the OP goal.

File: protocols/v2/binary-sv2/no-serde-sv2/codec/src/lib.rs
Comment (Shourya742) [inline protocols/v2/binary-sv2/no-serde-sv2/codec/src/lib.rs:23]: We can now use `#![no_std]`, and I'm confident that we are not using any `std`-compliant methods in `binary_codec_sv2`  in any of the top-level crates that depend on it. Since almost all of our crates are `no_std`-compliant, makes sense to have no-serde-sv2 to no-std only. cc: @rrybarczyk

**Review by Shourya742 — CHANGES_REQUESTED** [review]

File: protocols/v2/binary-sv2/no-serde-sv2/codec/src/lib.rs
Comment (Georges760) [reply protocols/v2/binary-sv2/no-serde-sv2/codec/src/lib.rs:23]: I asked about this in this [comment](https://github.com/stratum-mining/stratum/pull/1230#issuecomment-2441362158).

Current `binary_codec_sv2` has a broken `std` support, do we want to remove it or fix it ?

Look like you propose to remove it. It is also my preference.

File: protocols/v2/binary-sv2/binary-sv2/src/lib.rs
Comment (Georges760) [reply protocols/v2/binary-sv2/binary-sv2/src/lib.rs]: done

**Comment by plebhash** [comment]
> As far as I tried all protocol/v2/ crates (libs) can be 100% `no_std` a,d does not need the `std` (or `no_std`) feature to optionally add `std` dep.

if I understand correctly, we should expect this PR to leave all `protocols` crates (with exception of `noise_sv2` + `codec_sv2`, which will be covered in other PRs) without **ANY** `std`/`no_std` feature flag, since they will be `no_std` by default.

but I'm confused:

why are we still leaving some crates with the `no_std` feature flag? I inspected this branch and found the `no_std` feature flag on the following crates:
- `v2/codec-sv2/Cargo.toml` (expected, since this PR doesn't aim to touch this crate IIUC)
- `v2/subprotocols/job-declaration/Cargo.toml` 
- `v2/subprotocols/mining/Cargo.toml`
- `v2/binary-sv2/no-serde-sv2/codec/Cargo.toml`
- `v2/subprotocols/template-distribution/Cargo.toml`
- `v2/subprotocols/common-messages/Cargo.toml`

shouldn't `no_std` feature flag be completely removed from these crates?
shouldn't they simply have a `#![no_std]` on the top of the module, instead of `#![cfg_attr(feature = "no_std", no_std)]`?

**Comment by plebhash** [comment]
also, we should bump the `PATCH` version of the crates that are being changed

**Comment by Georges760** [comment]
> > As far as I tried all protocol/v2/ crates (libs) can be 100% `no_std` a,d does not need the `std` (or `no_std`) feature to optionally add `std` dep.
> 
> if I understand correctly, we should expect this PR to leave all `protocols` crates (with exception of `noise_sv2` + `codec_sv2`, which will be covered in other PRs) without **ANY** `std`/`no_std` feature flag, since they will be `no_std` by default.
> 
That is the end goal yes

> but I'm confused:
> 
> why are we still leaving some crates with the `no_std` feature flag? I inspected this branch and found the `no_std` feature flag on the following crates:
> 
> * `v2/codec-sv2/Cargo.toml` (expected, since this PR doesn't aim to touch this crate IIUC)
> * `v2/subprotocols/job-declaration/Cargo.toml`
> * `v2/subprotocols/mining/Cargo.toml`

I didn't consider subprotocols/* crates either in this PR

> * `v2/binary-sv2/no-serde-sv2/codec/Cargo.toml`

details [here](https://github.com/stratum-mining/stratum/pull/1230#issuecomment-2441362158) above
TL;DR: kept `no_std` feature for an existing yet broken implementation (don't want to remove any piece of code, even if doesn't compile, but yes I think we should also clean what is under the `#[cfg(not(feature = "no_std"))]` in this crate, maybe part of #1132 )

> * `v2/subprotocols/template-distribution/Cargo.toml`
> * `v2/subprotocols/common-messages/Cargo.toml`
> 
> shouldn't `no_std` feature flag be completely removed from these crates? shouldn't they KISS and only have a `#![no_std]` on the top of the module, instead of `#![cfg_attr(feature = "no_std", no_std)]`? e.g.: [422aee3](https://github.com/stratum-mining/stratum/commit/422aee3eecfd34877dc0363a92c5997b4cfb1ae7)

YES totally, but this PR being multi crate already, try to limit its action to not refactor any code, just remove `std` dependencies where possible *without* functional impact.

**Comment by Georges760** [comment]
> following [this criteria](https://github.com/stratum-mining/stratum/discussions/970#discussioncomment-10973798), we should bump the `PATCH` version of the crates that are being changed on this PR

exact, but usually the version bump is done in a separate PR/commit to merge multiple improvement in a single version increment. Should I do it in this PR anyway ?

File: protocols/v2/binary-sv2/no-serde-sv2/codec/src/lib.rs
Comment (rrybarczyk) [reply protocols/v2/binary-sv2/no-serde-sv2/codec/src/lib.rs:23]: I also think we should enforce `no_std` on the `protocols` crates. So let's remove.

Comment (Shourya742) [reply protocols/v2/binary-sv2/no-serde-sv2/codec/src/lib.rs:23]: This PR looks good to me and is ready to go. We can enforce the `no-std` restriction during our protocol refactoring effort, once we have more concrete reasoning for removing `std`-related imports. However, I’m fine with either approach.

**Review by Shourya742 — APPROVED** [review]
ACK

File: protocols/v2/binary-sv2/no-serde-sv2/codec/src/lib.rs
Comment (rrybarczyk) [reply protocols/v2/binary-sv2/no-serde-sv2/codec/src/lib.rs:23]: Agree. Let's confirm in the dev meet before we approve.

**Comment by plebhash** [comment]
> > following [this criteria](https://github.com/stratum-mining/stratum/discussions/970#discussioncomment-10973798), we should bump the `PATCH` version of the crates that are being changed on this PR
> 
> exact, but usually the version bump is done in a separate PR/commit to merge multiple improvement in a single version increment. Should I do it in this PR anyway ?

you're right, we can leave it for a new PR

**Review by rrybarczyk — APPROVED** [review]

**Comment by plebhash** [comment]
@Georges760 could you please add a comment to the crates `README.md` mentioning that they are `no_std` by default?

some crates are missing a `README.md` (WIP under #845)

for the crates that are missing a `README.md`, feel free to create the file and only add this sentence

later, we're going to finish the `README.md` file with the rest of the info

**Review by rrybarczyk — APPROVED** [review]

**Review by plebhash — APPROVED** [review]

**Pull request merged**
